sources/scraping/runner.py

```diff
@@ -0,0 +1,181 @@
+import threading
+import typing as t
+import dlt
+
+from dlt.common import logger
+from pydispatch import dispatcher  # type: ignore
+from typing_extensions import Self
+
+from scrapy import signals, Item, Spider  # type: ignore
+from scrapy.crawler import CrawlerProcess  # type: ignore
+
+from .types import AnyDict, Runnable, P
+from .queue import ScrapingQueue
+
+T = t.TypeVar("T")
+
+
+class Signals:
+    def __init__(self, pipeline_name: str, queue: ScrapingQueue[T]) -> None:
+        self.stopping = False
+        self.queue = queue
+        self.pipeline_name = pipeline_name
+
+    def on_item_scraped(self, item: Item) -> None:
+        if not self.queue.is_closed:
+            self.queue.put(item)
+        else:
+            logger.info(
+                "Queue is closed, stopping",
+                extra={"pipeline_name": self.pipeline_name},
+            )
+            self.stopping = True
+            self.crawler.stop()
+
+    def on_engine_stopped(self) -> None:
+        logger.info(f"Crawling engine stopped for pipeline={self.pipeline_name}")
+        self.stopping = True
+        self.queue.join()
+        self.queue.close()
+
+    def __call__(self, crawler: CrawlerProcess) -> Self:
```

Comment:
@rudolfix I added __call__ method here so the context manager expression with call looks better.